build cryptography=3.4.8 for python 3.10.*

fhoehle opened the pull request:
<!--
Thank you for pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] ~Bumped~ Changed the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->
Some packages like `snowflake-connector-python` restricting updates of `cryptography` by excluding the next major version i.e. `<4`. In order to progress with the `python=3.10` migration it would be helpful to build latest `cryptography=3.*` also for `python=3.10`. This PR could start some `abi_branch` if there are no objects. Please note: this PR **must not** be merged to master but to some different branch `3.4.x` maybe.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

fhoehle commented:
@conda-forge-admin, please rerender

h-vetinari commented:
@conda-forge/cryptography 
Do we want to do this?
If we do, I could create a branch for this, but given the infra changes since then (cross-compilation & various fixes), I'd actually suggest to branch off from current master rather than from before #71 was merged.

jakirkham commented:
@fhoehle can you please share why this is needed? This is quite an old version of a package used for...well...cryptography

h-vetinari commented:
> 
> 
> @fhoehle can you please share why this is needed? This is quite an old version of a package used for...well...cryptography

In his defense, 3.4.8 is [barely](https://github.com/pyca/cryptography/releases/tag/3.4.8) 3 months old (only that upstream changed their version scheme since then), and it's also the last version that doesn't hard-require rust support to build.

fhoehle commented:
> @fhoehle can you please share why this is needed? This is quite an old version of a package used for...well...cryptography

as @h-vetinari said it is not very old it is the second latest release from August 24 2021. Reasoning for is that a transition to `python=3.10` does not require a simultaneous transition from `cryptography=3.4.*` to `cryptography=35.*`.

@h-vetinari 
> If we do, I could create a branch for this, but given the infra changes since then (cross-compilation & various fixes), I'd actually suggest to branch off from current master rather than from before #71 was merged.

I branched off at master, see parent commit of this branch:
<img width="193" alt="grafik" src="https://user-images.githubusercontent.com/4427507/142660761-4f87a404-cbd5-479f-ae68-4a80dc37ba8c.png">

h-vetinari commented:
OK, I created a branch and pointed your PR to it. There's now a conflict to solve, and you'll have to remove the unconditional `skip: true` that I added so that the CI doesn't generate anything for the new branch yet.

The build number is already correct, there was only one [build](https://anaconda.org/conda-forge/cryptography/files?page=2) (`_0`) for 3.4.8.

fhoehle commented:
> OK, I created a branch and pointed your PR to it. There's now a conflict to solve, and you'll have to remove the unconditional `skip: true` that I added so that the CI doesn't generate anything for the new branch yet.

thanks @h-vetinari. I rebased on `3.4.x` and adapted the cross compiling patch to be applicable for `cryptography=3.4.8` updating the shipped Rust dependency `pyo3` from `0.13.1` to `0.15.0`. This fixes the Python 3.10 building problems on `osx-arm64`.

h-vetinari commented on recipe/patches/0001-Use-pyo3-0.15.0-to-fix-cross-compilation.patch:
```diff
@@ -1,25 +1,25 @@
-From c0aac7fbe675d3afcf5c64eed39f369d706f4fa1 Mon Sep 17 00:00:00 2001
-From: Chris Burr <christopher.burr@cern.ch>
-Date: Thu, 4 Nov 2021 08:33:08 +0100
+From 24584ff9e4f3b312108df4629852c3faecd1ae5f Mon Sep 17 00:00:00 2001
+From: fhoehle <fhoehle@users.noreply.github.com>
+Date: Sun, 21 Nov 2021 18:05:18 +0100
```
Please keep the attribution in the patch as it was.

beckermr replied:
I second this.

fhoehle replied:
Since the patch of `cryptography=35.0.0` was not applicable I generate one based on `cryptography=3.4.8` replacing the existing patch. But fine with me, to change the header, I do not think it is relevant for patching.

h-vetinari replied:
Rebasing is quite a different thing from coming up with it in the first place (which often contains lots of tries that end up not getting reflected in the final commit); so keeping that information is mostly about who to ask if something doesn't work anymore (and not having to go hunt for it in the git history).

h-vetinari commented:
I think this PR is fine to merge and poses no risk, but CC @conda-forge/core for visibility since we're going beyond what upstream does on a security-relevant package.

beckermr commented:
How did these two migrations become coupled?

h-vetinari commented:
> How did these two migrations become coupled?

`cryptography` doesn't have a migration. It's just that the 3.10 migrator came by after this feedstock had already been bumped to 3.4.8.

At first I thought that upstream only started supporting python 3.10 with v35.0.0 as well, but actually, they've (only) been [releasing](https://pypi.org/project/cryptography/3.4.8/#files) `abi3` wheels, so their 3.4.8 wheels can still be installed on 3.10 (just tested this).

h-vetinari reviewed (approved):
I think we should do this, but I'd like to have another approval from @conda-forge/cryptography or core.

To be clear: I was wrong in my original ping above: upstream supports 3.4.8 for python 3.10 by virtue of `abi3` wheels.

beckermr commented:
Cc @conda-forge/core any thoughts here?

xhochy commented:
👍 on this PR, I expect that we will see many more of these backports PRs for the 3.10 migration (like we did for past migrations).
